### PR TITLE
Infer from usage by similarity to builtin types

### PR DESCRIPTION
**Pull request by sandersn**
I think this PR is ready for review now.

This PR:
1. creates a list of "important" types. Right now: string, number, Array and Promise.
2. checks whether the members of an inferred type are assignable to any of the members of those types.
3. if 1 or 2 types match, these are added to the set of inferred types.

It also infers type parameters for Array and Promise, using an algorithm that is disturbingly close to the real inference algorithm. On the whole, infer-from-usage is like a shadow checker with subtly different rules. I don't really like this, but it may be inherent in the way infer-from-usage is designed. Thoughts?

Notes:

1. This needs a lot more tests. Edit: This is done; at least to a bare minimum.
2. There are three drive-by fixes:
  a. In `+` inference, `T + any` no longer infers `T = string | number`.
  b. inference type "unification" (it is more like aggregation and selection) no longer infers literal types and now uses subtype reduction when forming unions.
  c. Return type inference no longer defaults to `void`; being in an expression statement infers `void` instead.

I'm not sure (2a) is correct; #30093 tracks problems with `+` inference and has some examples I need to try.

Follow up work:

1. Better inference: I want to look at particular failures from the user test suite for ideas. Already I noticed that `concat` doesn't work because it has two overloads.
2. Inference from named types: I plan to scrape source files and the global symbol table for symbols with SymbolFlags.Class, then add them to the list of types that can be inferred.
3. Caching: inferFromUsage should be able to cache its results. I'm not doing it here because [cache invalidation is one of two hard things in CS](https://martinfowler.com/bliki/TwoHardThings.html).

None of these are particularly required but (2) should help a lot and not be too hard.

### CODE REVIEW

File: src/services/codefixes/inferFromUsage.ts [hunk 1]
```diff
@@ -401,34 +414,35 @@ namespace ts.codefix {
...
         interface CallUsage {
             argumentTypes: Type[];
-            returnType: Usage;
+            return_: Usage;
         }
 
         interface Usage {
-            isNumber?: boolean;
-            isString?: boolean;
+            isNumber: boolean | undefined;
```
Comment (sandersn) [inline src/services/codefixes/inferFromUsage.ts]: cleanup: all Usage properties are now required and Usages are created with createEmptyUsage.

File: src/services/codefixes/inferFromUsage.ts [hunk 2]
```diff
@@ -626,6 +640,9 @@ namespace ts.codefix {
                     else if (otherOperandType.flags & TypeFlags.StringLike) {
                         usage.isString = true;
                     }
+                    else if (otherOperandType.flags & TypeFlags.Any) {
```
Comment (sandersn) [inline src/services/codefixes/inferFromUsage.ts]: driveby fix (2a)

File: src/services/codefixes/inferFromUsage.ts [hunk 3]
```diff
@@ -772,7 +790,7 @@ namespace ts.codefix {
                 good = good.filter(i => !(checker.getObjectFlags(i) & ObjectFlags.Anonymous));
                 good.push(unifyAnonymousTypes(anons));
             }
-            return checker.getWidenedType(checker.getUnionType(good));
+            return checker.getWidenedType(checker.getUnionType(good.map(checker.getBaseTypeOfLiteralType), UnionReduction.Subtype));
```
Comment (sandersn) [inline src/services/codefixes/inferFromUsage.ts]: driveby fix (2b)

File: src/services/codefixes/inferFromUsage.ts [hunk 4]
```diff
@@ -875,45 +885,183 @@ namespace ts.codefix {
 
                 types.push(checker.createAnonymousType(/*symbol*/ undefined!, members, callSignatures, constructSignatures, stringIndexInfo, /*numberIndexInfo*/ undefined)); // TODO: GH#18217
             }
-            return types;
+            return types; // TODO: Should cache this since I HOPE it doesn't change
```
Comment (sandersn) [inline src/services/codefixes/inferFromUsage.ts]: like normal checking caching should be fine here. I'll do that in a followup PR.

File: src/services/codefixes/inferFromUsage.ts [hunk 5]
```diff
@@ -490,6 +551,9 @@ namespace ts.codefix {
             }
 
             switch (node.parent.kind) {
+                case SyntaxKind.ExpressionStatement:
+                    addCandidateType(usage, checker.getVoidType());
```
Comment (sandersn) [inline src/services/codefixes/inferFromUsage.ts]: driveby fix (2c)

File: src/services/codefixes/inferFromUsage.ts [hunk 6]
```diff
@@ -830,20 +901,23 @@ namespace ts.codefix {
             }
 
             types.push(...(usage.candidateTypes || []).map(t => checker.getBaseTypeOfLiteralType(t)));
+            types.push(...inferNamedTypesFromProperties(usage));
 
-            if (usage.properties && hasCalls(usage.properties.get("then" as __String))) {
-                const paramType = getParameterTypeFromCalls(0, usage.properties.get("then" as __String)!.calls!, /*isRestParameter*/ false)!; // TODO: GH#18217
-                const types = paramType.getCallSignatures().map(sig => sig.getReturnType());
```
Comment (sandersn) [inline src/services/codefixes/inferFromUsage.ts]: this was wrong--for the expression `numberPromise.then(n => n.toString())`, it would infer `Promise<string>` not `Promise<number>`.

**Comment by sandersn** [comment]
@andrewbranch @orta @weswigham I think this is ready for review now. Not sure who is most interested in this.

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -524,6 +524,9 @@ namespace ts {
             },
             getApparentType,
             getUnionType,
+            isTypeAssignableTo: (source, target) => {
```
Comment (weswigham) [inline src/compiler/checker.ts]: Even `@internal`, some people are going to be ecstatic that this finally found its way into our API surface and will therefore finally be callable by people using the API (even if via cast).

File: src/services/codefixes/inferFromUsage.ts [hunk 8]
```diff
@@ -834,92 +905,158 @@ namespace ts.codefix {
...
+                    return;
                 }
-
-                if (usage.constructs) {
-                    for (const construct of usage.constructs) {
-                        constructSignatures.push(getSignatureFromCall(construct));
-                    }
+                if (propUsage.calls) {
+                    const sigs = checker.getSignaturesOfType(source, SignatureKind.Call);
+                    result = result && !!sigs.length && checker.isTypeAssignableTo(source, getFunctionFromCalls(propUsage.calls));
```
Comment (andrewbranch) [inline src/services/codefixes/inferFromUsage.ts]: Here’s an interesting test case:

```ts
function test(a) {
  a.toString(2)
}
```

**Expected result**: `a` is `number`
**Actual result**: `a` is `{ toString: (arg0: number) => void; }`

It happens because every builtin’s `toString` call signature is assignable to the inferred structural one, and then the inference to builtins bails because of the limit of 3 union constituents. I briefly thought `source` and `target` might be backwards here—that you actually want to see if the _usage_ is assignable to the _builtin’s property_—but that’s too restrictive; the parameter type of `arg0: number` isn’t assignable to `radix?: number` under `strictFunctionTypes` because of the optionality of the latter, and the inferred return type `void` is not assignable to `string`. (Interestingly, if you swap `target` and `source`, turn off `strictFunctionTypes`, and change the line to `let x = a.toString(2)`, the inference to `number` occurs.)

I was able to hack up a fix for this something like:

```ts
const sigs = checker.getSignaturesOfType(source, SignatureKind.Call);
const usageType = getFunctionFromCalls(propUsage.calls);
const maxSourceParams = sigs.reduce((max, sig) => Math.max(max, sig.parameters.length), 0);
const maxUsageParams = usageType.getCallSignatures().reduce((max, sig) => Math.max(max, sig.parameters.length), 0);
result = result && !!sigs.length && maxUsageParams <= maxSourceParams && checker.isTypeAssignableTo(source, usageType);
```

but admittedly it feels kind of bad that checking assignability doesn’t Just Work™. Maybe I’m missing something clever?

File: src/services/codefixes/inferFromUsage.ts [hunk 9]
```diff
@@ -834,92 +905,158 @@ namespace ts.codefix {
...
+        }
 
-                if (usage.calls) {
-                    for (const call of usage.calls) {
-                        callSignatures.push(getSignatureFromCall(call));
-                    }
+        function allPropertiesAreAssignableToUsage(type: Type, usage: Usage) {
+            if (!usage.properties) return false;
+            let result = true;
+            usage.properties.forEach((propUsage, name) => {
```
Comment (andrewbranch) [inline src/services/codefixes/inferFromUsage.ts]: Small potatoes, but it seems like the `ts.forEachEntry` util would be useful here as it would allow you to terminate early after you see a result of `false`.

File: src/services/codefixes/inferFromUsage.ts (hunk 8 above)
Comment (weswigham) [reply src/services/codefixes/inferFromUsage.ts]: Errr, could we get better results by checking subtype first, then assignability, like we do for overload resolution?

File: src/services/codefixes/inferFromUsage.ts (hunk 9 above)
Comment (sandersn) [reply src/services/codefixes/inferFromUsage.ts]: Nice! I didn't know about that one.

File: src/services/codefixes/inferFromUsage.ts [hunk 10]
```diff
@@ -828,92 +899,158 @@ namespace ts.codefix {
...
 
-                types.push(checker.createAnonymousType(/*symbol*/ undefined!, members, callSignatures, constructSignatures, stringIndexInfo, /*numberIndexInfo*/ undefined)); // TODO: GH#18217
-            }
-            return types;
-
-            function recur(innerUsage: Usage): Type {
-                return unifyFromUsage(inferFromUsage(innerUsage));
+        /**
+         * inference is limited to
+         * 1. generic types with a single parameter
```
[the remark this answers is not in the document]
Comment (sandersn) [reply src/services/codefixes/inferFromUsage.ts]: Yep, plus I intend to expand this to other types later.

File: src/services/codefixes/inferFromUsage.ts (hunk 8 above)
Comment (sandersn) [reply src/services/codefixes/inferFromUsage.ts]: I think the root problem is that we expect to infer signatures whose lengths match, and have that beat the signatures whose lengths don't, right? I think an rule in inferFromUsage would be better than trying to get subtype or assignability to do the right thing, since they have other concerns to worry about -- they are part of the Real World Checker that has to be correct, but I feel like this preference is part of the Dream World Inference we have going on here, where you choose the best candidate on what feels right a lot of the time.

This would require maintaining some additional information -- specifically a list of source signatures matched with usage-signatures, and then that matched list would be used to rank the built-in types instead of just using `filter`. Basically, `combineNamedTypes`, by analogy with `combineTypes`. Maybe the decision could be even deferred to `combineTypes`.

Let's take this PR as-is and I'll think about it for the future.

Comment (weswigham) [reply src/services/codefixes/inferFromUsage.ts]: Oh, I mean, I just suggested subtype before assignability because i thought the difference in rules would already do the right thing (tm).

File: src/compiler/checker.ts (hunk 7 above)
Comment (SamVerschueren) [reply src/compiler/checker.ts]: I'm really interested in the reason why it should be marked as `@internal`. Don't get me wrong, I'm one of those people ecstatic that this found it's way to the API, now I can ditch my [custom typescript version](https://github.com/SamVerschueren/tsd/tree/master/libraries/typescript) from my library which exposed this. I'm just super curious about the reasoning behind this decision :)

Comment (orta) [reply src/compiler/checker.ts]: Effectively this function now has to cross conceptual boundaries in TypeScript. Going from being useful internally only inside checker to being used inside the TSServices, which meant it needed to be exposed as a public API on the checker internally within our codebase,

Comment (SamVerschueren) [reply src/compiler/checker.ts]: Ok, but why the `@internal` annotation? It looks like the method is still kinda private?

Comment (sandersn) [reply src/compiler/checker.ts]: I'm not 100% convinced that arbitrary uses of this function are safe and won't corrupt compiler state. Marking it `@internal` allows me to use it inside Typescript where I know how it's being used. Using it outside is an `any` cast away, but that way it's obvious that it's not technically supported.
